README.md
# Easy-Interface-Doc

快速生成 interface 文档工具, 针对后端 mock 出来的json 自动生成对应的类型.

```json
{
  "name": "小白",
  "age": 12,
  "isMarry": false
}
```

`=>`

```ts
interface Test {
  name: string;
  age: number;
  isMarry: boolean;
}

```



CHANGELOG.md
# 版本更新

`0.0.1`: 仅粗暴的支持整个选中 yapi 文档替换的情况。不支持递归。

`0.0.2`: 改回原有实现逻辑, 不支持最后一行转换

`0.0.3`: 支持任意选中替换

`0.1.1`: 正式版本, 但目前还不支持`undefined` , `null` 类型

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * 为ts申明文件提供便捷服务
 * 通过快捷键快速的将 yapi 上的接口
 * 改变成一个 interface 文件
 *
 * 1. 获取当前选中的文本
 * 2. 正则匹配选中的文本
 * 3. 写回原文本
 *
 * 参考:
 * - 选中
 * https://blog.csdn.net/Marksinoberg/article/details/89355341
 */
// import * as vscode from 'vscode';
const vscode = require('vscode');
const logic = require('./utils/logic');
function activate(context) {
    let disposable = vscode.commands.registerCommand('extension.changeTs', () => {
        const editor = vscode.editor || vscode.window.activeTextEditor; // 选中文件
        let editSelection = editor.selection; // 选择的内容
        editor.edit((editBuilder) => {
            let text = editor.document.getText(editSelection);
            let splitTpl = logic.parseInterface(text);
            // let splitTpl = logic.splitComment(text);
            replaceEditActiveStr(editBuilder, splitTpl);
        });
    });
    // 添加到命令栈
    context.subscriptions.push(disposable);
}
exports.activate = activate;
/**
 * 替换选中区域的文字
 * @param editBuilder 编辑对象
 * @param tpl 更改的字符模板
 */
function replaceEditActiveStr(editBuilder, tpl) {
    const { selection } = vscode.window.activeTextEditor;
    const { start, end } = selection;
    let reStart = new vscode.Position(start.line, start.character);
    const reEnd = new vscode.Position(end.line, end.character);
    editBuilder.replace(new vscode.Range(reStart, reEnd), tpl);
}
// this method is called when your extension is deactivated
function deactivate() {
    vscode.window.showInformationMessage('插件已卸载');
}
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/extension.ts
/*
 * @Author: LaoZhang
 * @Date: 2019-12-27 12:01:13
 * @LastEditors  : LaoZhange
 * @LastEditTime : 2020-01-15 17:39:19
 * @Description: 作用
 * @FilePath: /cli-demo/src/extension.ts
 */
import { TextEditor } from "vscode";

/**
 * 为ts申明文件提供便捷服务
 * 通过快捷键快速的将 yapi 上的接口
 * 改变成一个 interface 文件
 * 
 * 1. 获取当前选中的文本
 * 2. 正则匹配选中的文本
 * 3. 写回原文本
 * 
 * 参考:
 * - 选中
 * https://blog.csdn.net/Marksinoberg/article/details/89355341
 */
// import * as vscode from 'vscode';
const vscode = require('vscode');
const logic = require('./utils/logic');
export function activate(context: any) {

	let disposable = vscode.commands.registerCommand('extension.changeTs', () => {
		const editor = vscode.editor || vscode.window.activeTextEditor; // 选中文件
		let editSelection = editor.selection; // 选择的内容
		editor.edit((editBuilder: any) => { // 当编辑时调用
			let text = editor.document.getText(editSelection);
			let splitTpl = logic.parseInterface(text);
			// let splitTpl = logic.splitComment(text);
			replaceEditActiveStr(editBuilder, splitTpl)
		});
		
	});
	// 添加到命令栈
	context.subscriptions.push(disposable);
}
/**
 * 替换选中区域的文字
 * @param editBuilder 编辑对象
 * @param tpl 更改的字符模板
 */
function replaceEditActiveStr(editBuilder: any, tpl: string) {
	const {selection} = vscode.window.activeTextEditor;
	const { start , end } = selection;
	let reStart = new vscode.Position(start.line, start.character);
	const reEnd = new vscode.Position(end.line, end.character);
	editBuilder.replace(new vscode.Range(reStart, reEnd), tpl);
}
// this method is called when your extension is deactivated
export function deactivate() {
		vscode.window.showInformationMessage('插件已卸载');
}


out/utils/tools.js
"use strict";
/*
 * @Author: LaoZhang
 * @Date: 2020-01-15 09:11:49
 * @LastEditors  : LaoZhange
 * @LastEditTime : 2020-01-15 09:23:12
 * @Description: 作用
 * @FilePath: /easy-interface-doc/src/utils/tools.ts
 */
const firstUpperCase = (...rests) => {
    const [first, ...rest] = rests;
    first.toUpperCase() + rest.join('');
};
const is = function (type) {
    return function (target) {
        return ({}).toString.call(target) === `[object ${firstUpperCase(type)}]`;
    };
};
//  判断一个对象是否是数组
const isArray = Array.isArray || is('array');
const isString = is('string');
const isObject = is('object');
const isBoolean = is('boolean');
const isNumber = is('number');
const isJSON = (str) => {
    if (isObject(str))
        return true;
    if (!isString(str))
        return false;
    str = str.replace(/\s/g, '').replace(/\n|\r/, '');
    if (/^\{(.*?)\}$/.test(str))
        return /"(.*?)":(.*?)/g.test(str);
    if (/^\[(.*?)\]$/.test(str)) {
        return str.replace(/^\[/, '')
            .replace(/\]$/, '')
            .replace(/},{/g, '}\n{')
            .split(/\n/)
            .map(function (s) { return isJSON(s); })
            .reduce(function (prev, curr) { return !!curr; });
    }
    return false;
};
module.exports = {
    is,
    isArray,
    isObject,
    isString,
    isJSON,
    isBoolean,
    isNumber
};
//# sourceMappingURL=tools.js.map

src/utils/logic.ts
/*
 * @Author: LaoZhang
 * @Date: 2020-01-08 15:05:54
 * @LastEditors  : LaoZhange
 * @LastEditTime : 2020-01-19 16:36:49
 * @Description: 作用
 * @FilePath: /cli-demo/src/utils/logic.ts
 */
const tools = require('./tools');
/**
 * @description: 函数注释前面的长度
 * @param {Object} editor 当前激活文件
 * @return: lineSpace：前面的长度，frontStr：函数注释第一行的长度，line:当前行(数字)，nextLine 激活行的下一行是否有内容
 */
const lineSpaceFn = (editor: any) => {
  const line = editor.selection.active.line; // 当前行
  const lineProperty = editor.document.lineAt(line); // 当前行的属性
  let lineFirst = lineProperty.firstNonWhitespaceCharacterIndex; // 激活行 前面是否有值
  let lineSpace,
    nextLine,
    frontStr = ''; // 前面空几行
  const activeLine = editor.selection.active.line; // 激活行 行号
  // 判断当前行有没有内容 决定选择当前行还是下一行的长度
  if (
    lineProperty.isEmptyOrWhitespace &&
    editor._documentData.document.lineCount !== activeLine + 1
  ) {
    nextLine = activeLine + 1;
    lineSpace = editor.document.lineAt(nextLine)
      .firstNonWhitespaceCharacterIndex;
    lineFirst = lineFirst === 0 ? lineSpace : 0;
    frontStr = ''.padStart(lineFirst);
  } else {
    lineSpace = lineFirst;
  }
  return [lineSpace, frontStr, line, nextLine];
};

let regLine = /".*\n?/g;

/**
 * @description 将匹配的字符转换成 key: value 的格式, 暂不支持 undefined 和 null
 * @param text 
 * @returns `{"key": "value"}`
 */
const parseInterface = (text: string) => {
  let replaceText = text.replace(regLine, (str: string) => {
    let rpLine = str.replace(/"\w*" *: *(".*"|\d*),?/gi, (line: string) => {
      const [ key , value ] = line.split(':');
      let keyCode = determineKey(key);
      let valueCode = determineStringValueType(value);
      return `${keyCode}: ${valueCode}`
    });
    return rpLine;
  });
  return replaceText
}
/**
 * @description 转换 key
 * @param key 
 */
const determineKey = (key: string) => {
  return key.substring(1, key.length -1);
}
/**
 * @description 判断并转换对应类型的字符
 * @param value 
 */
const determineValueType = (type: any) => {
  // const { isObject, isArray, isString, isBoolean, isNumber } = tools;
  // let targetType: string;
  if (typeof type === 'object') {

  }
  return typeof type;
}
/**
 * @description 解析 JSON-value 的字符串所属类型
 * @param value 
 * @returns 返回类型
 */
const determineStringValueType = (str: string) => {
  // let str = String(value).trim().substring(0, value.length - 1)
  if (/"|'/g.test(str)) return 'string;'; // 有引号就代表是字符串类型
  if ( /\d/g.test(str) ) return 'number;';
  if ( /true|false/.test(str) ) return 'boolean;';
  if ( /undefined/.test(str) ) return 'undefined';
  if ( /null/.test(str) ) return 'null';
}

/**
 * @description JSON方法解析(弃用)
 * @param text 
 */
const splitComment = (text: string) => {
  const comments:  string[] = [];
  let replaceText = text.replace(/\/\/.*\n/g, (comment: string) => {
    comments.push(comment);
    return '';
  });
  const parseReplaceText = JSON.parse(replaceText);
  let result = `{\n`;
  Object.keys(parseReplaceText).map((key, idx: number) => {
    let currCommentVal = comments[idx];
    let value = parseReplaceText[key];
    let typeValue = `${determineValueType(value)}; ${currCommentVal}`; // 类型 + 注释
    result += `${key}: ${typeValue}`; // 简单类型
  });
  return `${result}}` 
}
module.exports = {
  lineSpaceFn,
  parseInterface,
  splitComment
};

src/utils/tools.ts
/*
 * @Author: LaoZhang
 * @Date: 2020-01-15 09:11:49
 * @LastEditors  : LaoZhange
 * @LastEditTime : 2020-01-15 09:23:12
 * @Description: 作用
 * @FilePath: /easy-interface-doc/src/utils/tools.ts
 */
const firstUpperCase = (...rests: string[]) => {
  const [ first, ...rest ] = rests;
  first.toUpperCase() + rest.join('')
};

const is = function (type: string) {
	return function (target: any) {
		return ({}).toString.call(target) === `[object ${firstUpperCase(type)}]`
	}
}
//  判断一个对象是否是数组
const isArray = Array.isArray || is('array')
const isString = is('string');
const isObject = is('object');
const isBoolean = is('boolean');
const isNumber = is('number');
const isJSON = (str: string): boolean => {
	if (isObject(str)) return true
	if (!isString(str)) return false;
	str = str.replace(/\s/g, '').replace(/\n|\r/, '');
	if (/^\{(.*?)\}$/.test(str))
		return /"(.*?)":(.*?)/g.test(str);
	if (/^\[(.*?)\]$/.test(str)) {
		return str.replace(/^\[/, '')
			.replace(/\]$/, '')
			.replace(/},{/g, '}\n{')
			.split(/\n/)
			.map(function (s) { return isJSON(s); })
			.reduce(function (prev, curr) { return !!curr; });
	}
	return false;
};
module.exports = {
  is,
  isArray,
  isObject,
  isString,
  isJSON,
  isBoolean,
  isNumber
};

out/utils/logic.js
"use strict";
/*
 * @Author: LaoZhang
 * @Date: 2020-01-08 15:05:54
 * @LastEditors  : LaoZhange
 * @LastEditTime : 2020-01-19 16:36:49
 * @Description: 作用
 * @FilePath: /cli-demo/src/utils/logic.ts
 */
const tools = require('./tools');
/**
 * @description: 函数注释前面的长度
 * @param {Object} editor 当前激活文件
 * @return: lineSpace：前面的长度，frontStr：函数注释第一行的长度，line:当前行(数字)，nextLine 激活行的下一行是否有内容
 */
const lineSpaceFn = (editor) => {
    const line = editor.selection.active.line; // 当前行
    const lineProperty = editor.document.lineAt(line); // 当前行的属性
    let lineFirst = lineProperty.firstNonWhitespaceCharacterIndex; // 激活行 前面是否有值
    let lineSpace, nextLine, frontStr = ''; // 前面空几行
    const activeLine = editor.selection.active.line; // 激活行 行号
    // 判断当前行有没有内容 决定选择当前行还是下一行的长度
    if (lineProperty.isEmptyOrWhitespace &&
        editor._documentData.document.lineCount !== activeLine + 1) {
        nextLine = activeLine + 1;
        lineSpace = editor.document.lineAt(nextLine)
            .firstNonWhitespaceCharacterIndex;
        lineFirst = lineFirst === 0 ? lineSpace : 0;
        frontStr = ''.padStart(lineFirst);
    }
    else {
        lineSpace = lineFirst;
    }
    return [lineSpace, frontStr, line, nextLine];
};
let regLine = /".*\n?/g;
/**
 * @description 将匹配的字符转换成 key: value 的格式, 暂不支持 undefined 和 null
 * @param text
 * @returns `{"key": "value"}`
 */
const parseInterface = (text) => {
    let replaceText = text.replace(regLine, (str) => {
        let rpLine = str.replace(/"\w*" *: *(".*"|\d*),?/gi, (line) => {
            const [key, value] = line.split(':');
            let keyCode = determineKey(key);
            let valueCode = determineStringValueType(value);
            return `${keyCode}: ${valueCode}`;
        });
        return rpLine;
    });
    return replaceText;
};
/**
 * @description 转换 key
 * @param key
 */
const determineKey = (key) => {
    return key.substring(1, key.length - 1);
};
/**
 * @description 判断并转换对应类型的字符
 * @param value
 */
const determineValueType = (type) => {
    // const { isObject, isArray, isString, isBoolean, isNumber } = tools;
    // let targetType: string;
    if (typeof type === 'object') {
    }
    return typeof type;
};
/**
 * @description 解析 JSON-value 的字符串所属类型
 * @param value
 * @returns 返回类型
 */
const determineStringValueType = (str) => {
    // let str = String(value).trim().substring(0, value.length - 1)
    if (/"|'/g.test(str))
        return 'string;'; // 有引号就代表是字符串类型
    if (/\d/g.test(str))
        return 'number;';
    if (/true|false/.test(str))
        return 'boolean;';
    if (/undefined/.test(str))
        return 'undefined';
    if (/null/.test(str))
        return 'null';
};
/**
 * @description JSON方法解析(弃用)
 * @param text
 */
const splitComment = (text) => {
    const comments = [];
    let replaceText = text.replace(/\/\/.*\n/g, (comment) => {
        comments.push(comment);
        return '';
    });
    const parseReplaceText = JSON.parse(replaceText);
    let result = `{\n`;
    Object.keys(parseReplaceText).map((key, idx) => {
        let currCommentVal = comments[idx];
        let value = parseReplaceText[key];
        let typeValue = `${determineValueType(value)}; ${currCommentVal}`; // 类型 + 注释
        result += `${key}: ${typeValue}`; // 简单类型
    });
    return `${result}}`;
};
module.exports = {
    lineSpaceFn,
    parseInterface,
    splitComment
};
//# sourceMappingURL=logic.js.map